content/src/content/jcr_root/apps/core/wcm/components/page/v2/page/dataLayer.html

```diff
@@ -0,0 +1,47 @@
+<!--/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright 2019 Adobe Systems Incorporated
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~     http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+
+<!--/* Data layer JavaScript object: */-->
+<script data-sly-use.page="com.adobe.cq.wcm.core.components.models.Page">
+    window.digitalData = window.digitalData || {};
+    window.digitalData.page = window.digitalData.page || {};
+    window.digitalData.page.attributes = window.digitalData.page.attributes || {};
+    window.digitalData.user = window.digitalData.user || {};
+    window.digitalData = {
+        page: {
+            pageInfo: {
+                pageName: '${currentPage.title @context="scriptString"}',
+                pageURL: window.location.href,
+                pageRef: document.referrer // Referring URL
+            	},
+            category:{
+            	primaryCategory: '${page.primaryCategory @context="scriptString"}', //Main Section as per the Site Hierarchy
+                subCategory: '${page.secondaryCategory @context="scriptString"}', //Site Section 1 as per the Site Hierarchy
+                pageType: '${page.templateName @context="scriptString"}'
+             	},
+            attributes:{
+            	locale: navigator.language, // Locale of the site
+                domain: window.location.hostname, // Domain of the site
+                language: '${page.language @context="scriptString"}', //Language where the user has visted(en, etc)
+                country: navigator.language.split("-")[1] //Country where the user has visted(US, etc)
+            	}
+        },
+        user: {
+            customerType: '${currentSession.userID @context="scriptString"}', // Logged in or anonymous

```

Comment:
Is there a caching risk here? i.e, if the page is cached at the dispatcher, would the first userid be cached?

Reply:
if the response headers are not explicitly cached at dispatcher, then there is no risk.

Reply:
From what I understand, this statement is part of the HTML, so **it is** getting cached. Isn't it possible to deduce this kind of information via javascript on client-side?

Reply:
If we are talking about CUGs being used here then there is anyway no caching. In most of the cases the end user login is not even known by AEM as this is transparently handled via some central IDP and client-side tokens. So I guess that should not be part of the data layer being provided by core components but rather would need to be added by custom implementation.

Reply:
+1 this must be implemented client side.

Reply:
Removed customerType attribute and moved script to js file following event driven data layer approach